### PR TITLE
fix: improve error message and option to skip unsupported transaction type in debug trace

### PR DESCRIPTION
**Pull request by agostbiro**
This PR resolves the following issues:

- https://github.com/NomicFoundation/edr/issues/597
- https://github.com/NomicFoundation/edr/issues/601

The approach taken is the one we discussed last Tuesday to fetch and filter transactions in the provider implementation of `debug_traceTransaction`. This is pretty hacky, but it has the advantage that changes are simple and localized and it'll be easy to remove once we have multi-chain support.

### CODE REVIEW

**Review by fvictorio — APPROVED** [review]
I tested this manually and confirmed that works as expected. I didn't review the code, so let's wait for @Wodann's approval.

File: crates/edr_provider/src/data.rs
Comment (Wodann) [inline crates/edr_provider/src/data.rs:2013]:
```suggestion
        // SAFETY: We only need the header from the block later, so it's safe to take the
```

I think it would be good to be explicit about this being related to safety. It makes it easier to search for and validate.

An unobservant editor might change the below code to return more than just the header of the remote block, breaking the "promise" that the `RemoteBlock` matches the received RPC block.

File: crates/edr_provider/src/data.rs
Comment (Wodann) [inline crates/edr_provider/src/data.rs:1923]:
```suggestion
        // HACK: This is a hack to make `debug_traceTransaction` return a helpful error
```

Nit: I find it helpful to have the `HACK:` prefix to make it easily searchable when looking for tech debt.

File: crates/edr_provider/tests/issues/issue_570.rs
Comment (Wodann) [inline crates/edr_provider/tests/issues/issue_570.rs:25]: This can be a `const`

File: crates/edr_provider/tests/issues/issue_570.rs
Comment (Wodann) [inline crates/edr_provider/tests/issues/issue_570.rs:48]: Why not compare the typed error? It seems like we have all information to make the test stronger

File: crates/edr_provider/tests/issues/issue_570.rs
Comment (Wodann) [inline crates/edr_provider/tests/issues/issue_570.rs:79]: It would be good to add a test that validates:
- `std::env::set_var("__EDR_UNSAFE_SKIP_UNSUPPORTED_TRANSACTION_TYPES", "true");`
- a debug trace where the `transaction_hash` is a transaction that's skipped
- assert that the error is `DebugTraceError::InvalidTransactionHash { transaction_hash, block_number: BLOCK_NUMBER }`

File: crates/edr_provider/src/data.rs
Comment (Wodann) [inline crates/edr_provider/src/data.rs:1949]: In `debug_trace_transaction`, we return `DebugTraceError::InvalidTransactionHash`. To retain the same behaviour, this should probably match that error.

The same applies for the occurrences below

**Review by Wodann — CHANGES_REQUESTED** [review]
Overall looks fine to me. I just highlighted a couple of things worth considering

File: crates/edr_provider/src/data.rs
Comment (agostbiro) [reply crates/edr_provider/src/data.rs:1949]: We returned `ProviderError::<LoggerErrorT>::InvalidTransactionHash` previously from `ProviderData::debug_trace_transaction` when a transaction was not found: https://github.com/NomicFoundation/edr/pull/606/files#diff-03b2ca6fd82ba7931a70b3b1ef672aa25b18c86bf643d21a930efa2691e701cfL618-L621

`DebugTraceError::InvalidTransactionHash` is only returned from `edr_evm::debug_trace_transaction` if the arguments passed to it from `ProviderData::debug_trace_transaction` are wrong.

So `DebugTraceError::InvalidTransactionHash` is returned in case of a bug in our codebase, while `ProviderError::<LoggerErrorT>::InvalidTransactionHash` is returned if the user supplies wrong data. This PR retains that behavior, so I think it's ok as is.

File: crates/edr_provider/src/data.rs
Comment (agostbiro) [reply crates/edr_provider/src/data.rs]: Fixed in https://github.com/NomicFoundation/edr/pull/606/commits/0e66f07f3fce6f2646fb479b538d003caf5ab95e

File: crates/edr_provider/src/data.rs
Comment (agostbiro) [reply crates/edr_provider/src/data.rs]: Fixed in https://github.com/NomicFoundation/edr/pull/606/commits/0e66f07f3fce6f2646fb479b538d003caf5ab95e

File: crates/edr_provider/tests/issues/issue_570.rs
Comment (agostbiro) [reply crates/edr_provider/tests/issues/issue_570.rs]: Fixed in: https://github.com/NomicFoundation/edr/pull/606/commits/0e66f07f3fce6f2646fb479b538d003caf5ab95e

File: crates/edr_provider/tests/issues/issue_570.rs
Comment (agostbiro) [reply crates/edr_provider/tests/issues/issue_570.rs]: I originally wanted to assert that the error message is the expected one to test things end-to-end, but that was too long so I cut it back. I agree that it's better to assert the enum than the short message. Fixed in https://github.com/NomicFoundation/edr/pull/606/commits/0e66f07f3fce6f2646fb479b538d003caf5ab95e

File: crates/edr_provider/tests/issues/issue_570.rs
Comment (agostbiro) [reply crates/edr_provider/tests/issues/issue_570.rs:84]: Thanks for catching this! I handled this slightly differently by adding a specific error for this case with a helpful message: `Cannot perform debug tracing on transaction '{transaction_hash:?}', because it has unsupported transaction type '{unsupported_transaction_type}'`. 

Commit with test: [84dccda](https://github.com/NomicFoundation/edr/pull/606/commits/84dccda418e7244e6c0b52f22903d2153e3f9fde)

File: crates/edr_provider/tests/issues/issue_570.rs
Comment (agostbiro) [reply crates/edr_provider/tests/issues/issue_570.rs]: Actually, really fixed it in [b11aac7](https://github.com/NomicFoundation/edr/pull/606/commits/b11aac74a3a5a53db664a29f0799acd64abec75a)

File: crates/edr_provider/src/data.rs
Comment (Wodann) [reply crates/edr_provider/src/data.rs:1949]: Thank you for explaining

**Review by Wodann — APPROVED** [review]

**Pull request merged**
